src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,392 @@
+        const LegacyControllerMapping::ScreenInfo& info,
+        ControllerScriptEngineBase* parent)
+        : QObject(),
+          m_screenInfo(info),
+          m_GLDataFormat(GL_RGBA),
+          m_GLDataType(GL_UNSIGNED_BYTE),
+          m_isValid(true),
+          m_pControllerEngine(parent) {
+    switch (m_screenInfo.pixelFormat) {
+    case QImage::Format_RGB16:
+        m_GLDataFormat = GL_RGB;
+        m_GLDataType = m_screenInfo.reversedColor
+                ? GL_UNSIGNED_SHORT_5_6_5_REV
+                : GL_UNSIGNED_SHORT_5_6_5;
+        break;
+    case QImage::Format_RGB888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGR : GL_RGB;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    case QImage::Format_RGBA8888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGRA : GL_RGBA;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    default:
+        m_isValid = false;
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pThread = std::make_unique<QThread>();
+    m_pThread->setObjectName("ControllerScreenRenderer");
+
+#if QT_VERSION >= QT_VERSION_CHECK(6, 7, 0)
+    [[maybe_unused]] bool successful = moveToThread(m_pThread.get());
+    DEBUG_ASSERT(successful);
+#else
+    moveToThread(m_pThread.get());
+#endif
+
+    // these at first sight weird-looking connections facilitate thread-safe communication.
+    connect(this,
+            &ControllerRenderingEngine::engineSetupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::sendFrameDataRequested,
+            this,
+            &ControllerRenderingEngine::send);
+    connect(m_pThread.get(),
+            &QThread::finished,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_pThread->start(QThread::NormalPriority);
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    DEBUG_ASSERT(QThread::currentThread() != thread());
+    m_pThread->wait();
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.critical() << "The ControllerEngine is being deleted but hasn't been "
+                              "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!thread()->isFinished() && !thread()->isInterruptionRequested()) {
+        kLogger.critical() << "Render thread has or is about to terminate. Cannot "
+                              "start this render anymore.";
+        return;
+    }
+    QCoreApplication::postEvent(this, new QEvent(QEvent::UpdateRequest));
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pThread && m_pThread->isRunning();
+}
+
+void ControllerRenderingEngine::requestEngineSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
```

Comment:
from `util/thread_affinity.h`
```suggestion
void ControllerRenderingEngine::requestEngineSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
    DEBUG_ASSERT_QOBJECT_THREAD_AFFINITY(m_pControllerEngine);
    m_isValid = false;
```
Technically don't know how safe that is as `QObject::thread()` is not marked thread-safe (so its only true if the assertion was true, but we can't know that beforehand 🫣), but I'm sure its fine. It's used all over mixxx.

Reply:
Added with `VERIFY..` and `ANTI_AFFINITY` variants